### ISSUE TITLE
Question: AutoMapper ExpressionMapping of a Select Expression with sub objects

### ISSUE THREAD

**Issue by Apolloangel** [AutoMapper/AutoMapper.Extensions.ExpressionMapping#136]
The issue I'm currently facing is that I have a Select expression based on a DTO model where some of the selected properties are part of a SubObject. I am rather sure this issue happens because I am doing wrong or at least I hope so so any guidance would be appreciated.

The issue happens when I try to map this expression:

```
.Lambda #Lambda1<System.Func`2[BoyumCloudFramework.OData.Test.OdataToIQueryableTest.HelperClasses.TestPilotDtoEntry,BoyumCloudFramework.OData.Test.OdataToIQueryableTest.HelperClasses.TestPilotDtoEntry]>(BoyumCloudFramework.OData.Test.OdataToIQueryableTest.HelperClasses.TestPilotDtoEntry $e)
{
    .New BoyumCloudFramework.OData.Test.OdataToIQueryableTest.HelperClasses.TestPilotDtoEntry(){
        Title = $e.Title,
        FirstName = $e.FirstName,
        TFlight = .New BoyumCloudFramework.OData.Test.OdataToIQueryableTest.HelperClasses.TestFlightDto(){
            Destination = ($e.TFlight).Destination
        },
        Key = $e.Key
    }
}
```

To a DB expression
I get the following error:

```
AutoMapper.AutoMapperMappingException: Error mapping types.

Mapping types:
Expression1`1 -> Expression`1
System.Linq.Expressions.Expression1`1[[System.Func`2[[BoyumCloudFramework.OData.Test.OdataToIQueryableTest.HelperClasses.TestPilotDtoEntry, BoyumCloudFramework.WebApi.OData.Test, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null],[BoyumCloudFramework.OData.Test.OdataToIQueryableTest.HelperClasses.TestPilotDtoEntry, BoyumCloudFramework.WebApi.OData.Test, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]], System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]] -> System.Linq.Expressions.Expression`1[[System.Func`2[[BoyumCloudFramework.OData.Test.OdataToIQueryableTest.HelperClasses.TestPilotDbEntry, BoyumCloudFramework.WebApi.OData.Test, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null],[BoyumCloudFramework.OData.Test.OdataToIQueryableTest.HelperClasses.TestPilotDbEntry, BoyumCloudFramework.WebApi.OData.Test, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]], System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]
 ---> System.InvalidOperationException: No coercion operator is defined between types 'BoyumCloudFramework.OData.Test.OdataToIQueryableTest.HelperClasses.TestFlightDto' and 'BoyumCloudFramework.OData.Test.OdataToIQueryableTest.HelperClasses.TestFlightDb'.
   at System.Linq.Expressions.Expression.GetUserDefinedCoercionOrThrow(ExpressionType coercionType, Expression expression, Type convertToType)
   at System.Linq.Expressions.Expression.Convert(Expression expression, Type type, MethodInfo method)
   at System.Linq.Expressions.Expression.Convert(Expression expression, Type type)
   at AutoMapper.Extensions.ExpressionMapping.Extensions.VisitorExtensions.ConvertTypeIfNecessary(Expression expression, Type memberType)
   at AutoMapper.Extensions.ExpressionMapping.XpressionMapperVisitor.DoBind(MemberInfo sourceMember, Expression initial, Expression mapped)
   at AutoMapper.Extensions.ExpressionMapping.XpressionMapperVisitor.<>c__DisplayClass27_0.<GetMemberInit>b__0(List`1 list, MemberAssignmentInfo next)
   at System.Linq.Enumerable.Aggregate[TSource,TAccumulate](IEnumerable`1 source, TAccumulate seed, Func`3 func)
   at AutoMapper.Extensions.ExpressionMapping.XpressionMapperVisitor.GetMemberInit(MemberBindingGroup memberBindingGroup)
   at AutoMapper.Extensions.ExpressionMapping.XpressionMapperVisitor.VisitMemberInit(MemberInitExpression node)
   at System.Linq.Expressions.MemberInitExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at AutoMapper.Extensions.ExpressionMapping.MapperExtensions.<MapExpression>g__MapBody|8_1[TDestDelegate](Dictionary`2 typeMappings, XpressionMapperVisitor visitor, <>c__DisplayClass8_0`1& )
   at AutoMapper.Extensions.ExpressionMapping.MapperExtensions.<MapExpression>g__CreateVisitor|8_0[TDestDelegate](Dictionary`2 typeMappings, <>c__DisplayClass8_0`1& )
   at AutoMapper.Extensions.ExpressionMapping.MapperExtensions.MapExpression[TDestDelegate](IConfigurationProvider configurationProvider, LambdaExpression expression, Type typeSourceFunc, Type typeDestFunc, Func`3 getVisitor, Func`2 shouldConvertMappedBodyToDestType)
   at AutoMapper.Extensions.ExpressionMapping.MapperExtensions.MapExpression[TDestDelegate](IMapper mapper, LambdaExpression expression, Func`3 getVisitor, Func`2 shouldConvertMappedBodyToDestType)
   at AutoMapper.Extensions.ExpressionMapping.MapperExtensions.MapExpression[TDestDelegate](IMapper mapper, LambdaExpression expression)
   at AutoMapper.Mappers.ExpressionMapper.Map[TSource,TDestination](TSource expression, IConfigurationProvider configurationProvider)
   at lambda_method234(Closure , Object , Expression`1 , ResolutionContext )
   --- End of inner exception stack trace ---
   at lambda_method234(Closure , Object , Expression`1 , ResolutionContext )
   at BoyumCloudFramework.OData.Test.OdataCosmosOrchestratorTest.SelectExpressionMapping() in C:\git\BoyumCloudFramework\WebApi\OData.Test\OdataCosmosOrchestratorTest.cs:line 589
```

I have the following setup

```
[DataContract]
public class TestPilotDtoEntry : BaseDtoEntry
{
    [DataMember(Name = "Title")]
    public string? Title { get; set; }

    [DataMember(Name = "Fn")]
    public string? FirstName { get; set; }

    [DataMember(Name = "Ln")]
    public string? LastName { get; set; }

    [DataMember(Name = "Tf")]
    public TestFlightDto? TFlight { get; set; }
}
[DataContract]
public class TestFlightDto
{
    [DataMember(Name = "D")]
    public string? Destination { get; set; }
}

public class TestPilotDbEntry : BaseDbEntry
{
    [JsonProperty("T")]
    public string? Title { get; set; }

    [JsonProperty("F")]
    public string? FirstName { get; set; }

    [JsonProperty("L")]
    public string? LastName { get; set; }

    [JsonProperty("Tf")]
    public TestFlightDb? TestF { get; set; }
}

public class TestFlightDb
{
    [JsonProperty("D")]
    public string? Destination { get; set; }
}
```

\*\*Profile \*\*

```
public class TestPilotProfile : BaseProfile
{
    public TestPilotProfile()
    {
        CreateDbMapWithDefaultProperties<TestPilotDbEntry, TestPilotDtoEntry>()
            .ForMember(dest => dest.FirstName, opt =>
            {
                opt.MapFrom(src => src.FirstName);
                opt.ExplicitExpansion();
            })
            .ForMember(dest => dest.LastName, opt =>
            {
                opt.MapFrom(src => src.LastName);
                opt.ExplicitExpansion();
            })
            .ForMember(dest => dest.TFlight, opt =>
            {
                opt.MapFrom(src => src.TestF);
            })
            .ForMember(dest => dest.Title, opt =>
            {
                opt.MapFrom(src => src.Title);
                opt.ExplicitExpansion();
            });

       

        CreateDtoMapWithIgnoredProperties<TestPilotDtoEntry, TestPilotDbEntry>()
            .ForMember(dest => dest.FirstName, opt =>
            {
                opt.MapFrom(src => src.FirstName);
                opt.ExplicitExpansion();
            })
            .ForMember(dest => dest.LastName, opt => opt.MapFrom(src => src.LastName))
            .ForMember(dest => dest.TestF, opt => opt.MapFrom(src => src.TFlight))
            .ForMember(dest => dest.Title, opt => opt.MapFrom(src => src.Title));



        CreateMap<TestFlightDto, TestFlightDb>().ForMember(dest => dest.Destination, opt => opt.MapFrom(src => src.Destination));
        CreateMap<TestFlightDb, TestFlightDto>().ForMember(dest => dest.Destination, opt => opt.MapFrom(src => src.Destination));
    }
}
```

Issue happens when i run the following lines of code:

```
        var dtoSelectExpression = Helpers.BuildDtoSelectExpression<TestPilotDtoEntry, TestPilotDtoEntry>(membersToSelect);
            var dbSelectExpression = _mapper.Map<Expression<Func<TestPilotDbEntry, TestPilotDbEntry>>>(dtoSelectExpression);
```

**Comment by BlaiseD** [comment]
Can this be executed? (`CreateDbMapWithDefaultProperties` looks custom).  Please post a snippet we can run and reproduce.

**Comment by Apolloangel** [comment]
As requested here is CreateDbMapWithDefaultProperties. the method is responsible for mapping the properties of 2 base classes the Dto and Db  inherit from here is the method plus the two base models I'm using. 


```
public class BaseProfile : Profile
{
    /// <summary>
    /// Initializes a new instance of the <see cref="BaseProfile"/>.
    /// </summary>
    public BaseProfile()
    {
    }


    /// <summary>
    ///  A base map from a  <see cref="BaseDbEntry"/> to  <see cref="BaseDtoEntry"/>
    /// </summary>
    /// <typeparam name="TDb">A DbEntry inheriting <see cref="BaseDbEntry"/></typeparam>
    /// <typeparam name="TDto">A DtoEntry inheriting <see cref="BaseDtoEntry"/></typeparam>
    /// <returns></returns>
    protected IMappingExpression<TDb, TDto> CreateDbMapWithDefaultProperties<TDb, TDto>() where TDto : BaseDtoEntry where TDb : BaseDbEntry
    {
        return CreateMap<TDb, TDto>()
            .ForMember(dest => dest.Key, opt =>
            {
                opt.MapFrom(src => src.Key);
                opt.ExplicitExpansion();
            })
            .ForMember(dest => dest.ETag, opt => opt.MapFrom(src => src.ETag))
            .ForMember(dest => dest.Type, opt => opt.MapFrom(src => src.Type))
            .ForMember(dest => dest.MarkedDeleted, opt =>
            {
                opt.MapFrom(src => src.MarkedDeleted);
                opt.ExplicitExpansion();
            })
            .ForMember(dest => dest.UpdatedBy, opt =>
            {
                opt.MapFrom(src => src.UpdatedBy);
                opt.ExplicitExpansion();
            })
            .ForMember(dest => dest.LastUpdated, opt =>
            {
                opt.MapFrom(src => src.LastUpdated);
                opt.ExplicitExpansion();
            });
        
    }

    /// <summary>
    ///  A base map from a <see cref="BaseDtoEntry"/> to <see cref="BaseDbEntry"/> ignoring specific properties
    /// </summary>
    /// <typeparam name="TDto">A DtoEntry inheriting <see cref="BaseDtoEntry"/></typeparam>
    /// <typeparam name="TDb">A DbEntry inheriting <see cref="BaseDbEntry"/></typeparam>
    /// <returns></returns>
    protected IMappingExpression<TDto, TDb> CreateDtoMapWithIgnoredProperties<TDto, TDb>() where TDto : BaseDtoEntry where TDb : BaseDbEntry
    {
        return CreateMap<TDto, TDb>()
            .ForMember(dest => dest.ETag, opt => opt.MapFrom(src => src.ETag))
            .ForMember(dest => dest.Key, opt => opt.MapFrom(src => src.Key))
            .ForMember(dest => dest.LastUpdated, opt => opt.Ignore())
            .ForMember(dest => dest.UniqueIdentifier, opt => opt.Ignore())
            .ForMember(dest => dest.DataStoreId, opt => opt.Ignore())
            .ForMember(dest => dest.UpdatedBy, opt => opt.Ignore())
            .ForMember(dest => dest.MarkedDeleted, opt => opt.Ignore())
            .ForMember(dest => dest.Type, opt => opt.Ignore());
    }
}


```

```
public abstract class BaseDtoEntry
{
    /// <summary>
    ///     The key of the entry. This is unique
    /// </summary>
    [Key]
    [DataMember(Name = "Key")]
    public string? Key { get; set; }

    /// <summary>
    ///     Used for optimistic concurrency
    /// </summary>
    [DataMember(Name = "ETag")]
    public string? ETag { get; set; }

    /// <summary>
    ///     The last time entry was updated
    /// </summary>
    [DataMember(Name = "LastUpdated")]
    public DateTimeOffset? LastUpdated { get; set; }

    /// <summary>
    ///     The last entity to update the entry
    /// </summary>
    [DataMember(Name = "UpdatedBy")]
    public string? UpdatedBy { get; set; }

    /// <summary>
    ///     Type of the entry in the Database
    /// </summary>
    [DataMember(Name = "Type")]
    public string? Type { get; set; }

    /// <summary>
    ///     is the entry marked as deleted
    /// </summary>
    [DataMember(Name = "MarkedDeleted")]
    public bool? MarkedDeleted { get; set; }
}

public abstract class BaseDbEntry
{
    /// <summary>
    ///     What company does this object belong to
    /// </summary>
    [JsonProperty("DsId")]
    public string DataStoreId { get; set; } = string.Empty;

    /// <summary>
    ///     Unique id pr. datastore. This is a combination key of Type and Key.
    /// </summary>
    [JsonProperty("id")]
    public string UniqueIdentifier { get; set; } = string.Empty;

    /// <summary>
    ///     The key of the entry. This is unique pr. type and datastore. This is used for auto generated keys
    /// </summary>
    [JsonProperty("Key", Required = Required.Always)]
    public string Key { get; set; } = Guid.NewGuid().ToString();

    /// <summary>
    ///     The type of the object. This is automatically set by the ObjectGenerator. No need to set it yourself.
    /// </summary>
    [JsonProperty("_TY")]
    public string Type { get; set; } = string.Empty;

    /// <summary>
    ///     Used for optimistic concurrency
    /// </summary>
    [JsonProperty("_etag")]
    public string ETag { get; set; } = string.Empty;

    /// <summary>
    ///     Last Updated. This is automatically set. No need to set it yourself.
    /// </summary>
    [JsonProperty("_LU")]
    public DateTimeOffset LastUpdated { get; set; }

    /// <summary>
    ///     Updated by. This is automatically set. No need to set it yourself.
    /// </summary>
    [JsonProperty("_UB")]
    public virtual string UpdatedBy { get; set; } = string.Empty;

    /// <summary>
    ///     True if the document has been marked for deletion
    /// </summary>
    [JsonProperty("_DEL")]
    public bool MarkedDeleted { get; set; }
}

```
About: 

> I think this
> 
> ```
> defined between types 
> 'BoyumCloudFramework.OData.Test.OdataToIQueryableTest.HelperClasses.TestFlightDto' and 
> 'BoyumCloudFramework.OData.Test.OdataToIQueryableTest.HelperClasses.TestFlightDb'
> ```
> 
> is saying part of the expression is not getting mapped.

I totally agree that this seems to be the issue just don't know why it is not being mapped, since I have created a map between the two models.

**Comment by BlaiseD** [comment]
You're still missing pieces need to run the code:
```c#
            var dtoSelectExpression = Helpers.BuildDtoSelectExpression<TestPilotDtoEntry, TestPilotDtoEntry>(membersToSelect);
            var dbSelectExpression = _mapper.Map<Expression<Func<TestPilotDbEntry, TestPilotDbEntry>>>(dtoSelectExpression);
```

Your report should resemble the following:
```c#
    internal class Program
    {
        static void Main()
        {
            var config = new MapperConfiguration(cfg =>
            {
                cfg.AddExpressionMapping();
                cfg.AddMaps(typeof(TestPilotProfile));
            });

            IMapper mapper = config.CreateMapper();

            Expression<Func<TestPilotDtoEntry, TestPilotDtoEntry>> dtoSelectExpression = **_someExpression_**;
            var dbSelectExpression = mapper.Map<Expression<Func<TestPilotDbEntry, TestPilotDbEntry>>>(dtoSelectExpression);
        }
    }

//Add supporting classes here
```

- Anyone trying to help should be able to copy, paste, run, and reproduce the error (maybe add using statements).
- Please leave out attributes like `JsonProperty` and `DataContract` which are not relevant to the problem.

I'll close this one because the info is a little scattered given the back and forth.   If you still need help please open a new issue with all the code together as described.

Make sense?

**Comment by BlaiseD** [comment]
Resolved in issue #137 and  PR  #138.
